---
title: "Mutational signatures impact the evolution of anti-EGFR antibody resistance in colorectal cancer"
subtitle: "Figure S7"
author: "Andrew Woolston"
date: "06/01/2021"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

The following script reproduces Figure S7 in the publication 'Mutational signatures impact the evolution of anti-EGFR antibody resistance in colorectal cancer'.

# Load Libraries

```{r load required libraries,message=FALSE}

library(devtools)
library(readxl)
library(GenomicRanges)
library(SigsPack)
library(BSgenome.Hsapiens.UCSC.hg19)

```

# External documents

We do not have the permission to share all of the external data required to fully run this script. However, the data can be be accessed from the source locations indicated.

```{r external files,echo=T,eval=T}

## Load the Sigprofiler exome reference signatures from the PCAWG7 synapse repository (Synapse ID - syn12026190).
cosmic_filename<-'COSMIC_Mutational_Signatures_v3.1_SBS_GRCh37_WXS.xlsx'

## SureSelect V5 regions file
regions_filename<-'SSV5_regions.bed'

```

# Patients and signatures of interest

The model is used to assess how pre-treatment signature activity may influence the generation of specific KRAS/NRAS driver mutations at acquired resistance. We therefore use the signature exposures of baseline tumours (pre- cetuximab treatment) with prolonged benefit to create the mutation probability profile. 

```{r,echo=T,eval=T}

##BL tumours with prolonged benefit
selected.cases<-c('C1005BL', 'C1007BL', 'C1014BL', 'C1018BL', 'C1020BL', 'C1024BL', 'C1025BL', 'C1026BL', 'C1027BL', 'C1030BL', 'C1037BL', 'C1044BL')

## Relevent COSMIC signatures - see the manuscript for a description of how these signatures were selected.
selected.signatures<- c('SBS1','SBS3','SBS5','SBS6','SBS15','SBS17a','SBS17b','SBS35','SBS40')

```

# Normalise reference signatures

Mutation signature profiles are influenced by the frequency of each trinucleotide context across the exome. For instance, if a specific context is more abundant, then finding a mutation of that context will in turn be more likely than a lesser occuring motif (all other factors being equal). However, our modelling requires the probability of generating a specific hotspot mutation among hotspots. Therefore, we must adjust the canonical signatures to represent an even frequency of the 96 mutation motifs to assess the likelihood of generating that mutation. 

To do this we use the normalize function in the SigsPack R library. This is designed to convert mutation catalogues or reference signatures from a source distibution to a target distribution. It is typically used to convert signatures from exome to genome (or vice-versa). 

We are using reference signatures derived exclusively from WXS samples and so the conversion should be exome frequency to even distibution. Here we use the normalize function to run that conversion. 

```{r,echo=T,eval=T}

## Load the canonical signature profiles
Canonical_Signatures <- as.matrix(read_excel(cosmic_filename)); 

## set rownames to common trinucleotide mutation context notation
rownames(Canonical_Signatures)<-paste0(substr(Canonical_Signatures[,'SubType'],1,1),'[',Canonical_Signatures[,'Type'],']',substr(Canonical_Signatures[,'SubType'],3,3));

## extract signatures of interest
Canonical_Signatures_selected<-t(Canonical_Signatures)[selected.signatures,]

## convert to a numeric object
class(Canonical_Signatures_selected)<-"numeric"

## Load the exome regions file used for sequencing.
regions <- read.delim(regions_filename, header=FALSE, stringsAsFactors=FALSE)

## convert to Granges object
gr<-GenomicRanges::GRanges(seqnames=regions[,1],ranges=IRanges::IRanges(start=regions[,2],end=regions[,3]),strand=c("*"))

## gather all trinucleotide contexts across the exome
exome_context <- SigsPack::get_context_freq(BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19, gr)

## set the source and target conversion frequencies
source_context<-target_context<-exome_context
target_context[,1]<-sum(source_context)/length(source_context)

## run the normalize function to adjuste the reference signatures
Canonical_Signatures_selected_ExomeScaled<-SigsPack::normalize(t(Canonical_Signatures_selected), source_context = source_context, target_context = target_context)

```

# Baseline prolonged benefit signature proportions

We now use the signature attributions from the 12 baseline prolonged benefit tumours to calculate signature proportions representative of the group. The signature weights have been presented as Supplementary Table S2 in the manuscript. 

```{r,echo=T,eval=T}

## Load signature data and assign signature labels as rownames (see Supplementary Table S2)
Signature.Matrix <- as.matrix(read_excel("Supplementary_Table_S2.xlsx")); rownames(Signature.Matrix)<-Signature.Matrix[,'Signatures'];

## Extract assigned signature weights and convert to numeric
Signature.Weights<-Signature.Matrix[selected.signatures,][,selected.cases]
class(Signature.Weights)<-"numeric"

## Extract the explained variance
R2<-as.numeric(Signature.Matrix['R2',][selected.cases]); names(R2)<-selected.cases;

## Extract the mutation load of each sample
Mutation.Load<-as.numeric(Signature.Matrix['MutationLoad',][selected.cases]); names(Mutation.Load)<-selected.cases;

## Adjust signature weights by the R2 value
Signature.Weights.mat.adjusted<-t(t(Signature.Weights) * R2)

## Calculate the mutation load attributed to each signature
Signature.AssignedMutations<-rbind(t(t(Signature.Weights.mat.adjusted) * Mutation.Load),"unexplained proportion"=(1-R2)*Mutation.Load);

## Sum the mutations attributed to each signature across the 12 tumors
BL.ProlongedBenefit.Summary<-rowSums(Signature.AssignedMutations[,selected.cases])

## Calculate the overall proportional signature contributions
prop.sig.contribution<-(rowSums(Signature.AssignedMutations)/sum(BL.ProlongedBenefit.Summary))[selected.signatures]

## Scale the signature constributions to sum to 1 (exclude the unexplained)
prop.sig.contribution.nomiss<-prop.sig.contribution*(1/sum(prop.sig.contribution))

```

# Figure S7A

Figure S7A is a mutation probability profile generated using the signature proportions calculated above. We multiply the adjusted canonical signatures by the signature proportions and plot the probabilities indicating each signature contribution. 

```{r,echo=T,eval=T,fig.width=10, fig.height=5, fig.align = "center"}

## weight the signature contributions by observed proportions
select.default.weighted<-as.matrix(prop.sig.contribution.nomiss*t(Canonical_Signatures_selected_ExomeScaled)); class(select.default.weighted)<-"numeric";

colours<-rbind(c("SBS1","lightblue"),
               c("SBS3","darkblue"),
               c("SBS5","lightgreen"),
               c("SBS6","yellow"),
               c("SBS15","darkgreen"),
               c("SBS17a","orangered"),
               c("SBS17b","red"),
               c("SBS35","purple"),
               c("SBS40","orange"))

colour_code<-colours[,2][match(rownames(select.default.weighted),colours[,1])]

## plot modelled mutation profile
barplot(select.default.weighted,col=colour_code,las=2,cex.names=0.5)

```

# Figure S7B

Here we generate the proportional difference in KRAS/NRAS Q61H mutations versus all other observed G12/G13/Q61 mutations. We extract the relevent mutations from the Prospect and Bettegowda studies to calculate these proportions. 

```{r reported hotspot mutations,echo=F,eval=T}

## Prospect C reported mutations (Khan et al., Cancer Discovery, 2018)
Prospect1_KnownMuts<-rbind(c("chr12","25398285","25398285","C","A","C1005","KRAS","G12C"), ## KRAS G12C
                        c("chr1","115256530","115256530","G","T","C1005","NRAS","Q61K"), ## NRAS Q61K
                        c("chr12","25398285","25398285","C","A","C1044","KRAS","G12C"),  ## KRAS G12C
                        c("chr12","25380275","25380275","T","G","C1044","KRAS","Q61H"),  ## KRAS Q61H
                        c("chr1","115256530","115256530","G","T","C1044","NRAS","Q61K") ## NRAS Q61K
)

## Prospect C mutations (Woolston et al, Cancer Cell, 2019)
Prospect2_KnownMuts<-rbind(c("chr1","115256530","115256530","G","T","C1041","NRAS","Q61K"), ## NRAS Q61K
                          c("chr12","25398284","25398284","C","A","C1030","KRAS","G12V"),  ## KRAS G12V
                          c("chr12","25398284","25398284","C","T","C1030","KRAS","G12D"),  ## KRAS G12D
                          c("chr12","25398281","25398281","C","T","C1030","KRAS","G13D"),  ## KRAS G13D
                          c("chr12","25380275","25380275","T","G","C1025","KRAS","Q61H"),  ## KRAS Q61H
                          c("chr12","25380275","25380275","T","G","C1030","KRAS","Q61H"),  ## KRAS Q61H
                          c("chr12","25380275","25380275","T","G","C1030","KRAS","Q61H"),  ## KRAS Q61H
                          c("chr12","25380275","25380275","T","A","C1030","KRAS","Q61H"),  ## KRAS Q61H
                          c("chr12","25380275","25380275","T","G","C1037","KRAS","Q61H"))  ## KRAS Q61H
                          
## Detection of circulating tumor ... (Bettegowda et al., Science Translational Medicine, 2014)
Bettegowda_KnownMuts<-rbind(c("chr12","25398284","25398284","C","T","Patient5","KRAS","G12D"),  ## KRAS G12D
                            c("chr12","25380275","25380275","T","A","Patient16","KRAS","Q61H"), ## KRAS Q61H
                            c("chr12","25398284","25398284","C","A","Patient18","KRAS","G12V"), ## KRAS G12V
                            c("chr1","115258748","115258748","C","T","Patient18","NRAS","G12S"), ## NRAS G12S
                            c("chr1","115256528","115256528","T","A","Patient19","NRAS","Q61H"), ## NRAS Q61H
                            c("chr12","25398284","25398284","C","A","Patient21","KRAS","G12V"), ## KRAS G12V
                            c("chr1","115256528","115256528","T","A","Patient21","NRAS","Q61H"), ## NRAS Q61H
                            c("chr1","115256529","115256529","T","C","Patient21","NRAS","Q61R"), ## NRAS Q61R
                            c("chr12","25398284","25398284","C","G","Patient22","KRAS","G12A"), ## KRAS G12A
                            c("chr12","25398285","25398285","C","A","Patient22","KRAS","G12C"), ## KRAS G12C
                            c("chr12","25398285","25398285","C","T","Patient22","KRAS","G12S"), ## KRAS G12S
                            c("chr12","25398284","25398284","C","T","Patient22","KRAS","G12D"), ## KRAS G12D
                            c("chr12","25398284","25398284","C","A","Patient22","KRAS","G12V"), ## KRAS G12V
                            c("chr12","25398284","25398284","C","G","Patient24","KRAS","G12A"), ## KRAS G12A
                            c("chr1","115256530","115256530","G","T","Patient24","NRAS","Q61K"), ## NRAS Q61K
                            c("chr12","25398284","25398284","C","A","Patient26","KRAS","G12V"), ## KRAS G12V
                            c("chr12","25398284","25398284","C","A","Patient26","KRAS","G12V"), ## KRAS G12V
                            c("chr12","25398284","25398284","C","G","Patient1","KRAS","G12A"), ## KRAS G12A
                            c("chr12","25398285","25398285","C","A","Patient1","KRAS","G12C"), ## KRAS G12C
                            c("chr12","25398285","25398285","C","G","Patient1","KRAS","G12R"), ## KRAS G12R
                            c("chr12","25398284","25398284","C","T","Patient1","KRAS","G12D"), ## KRAS G12D
                            c("chr12","25398284","25398284","C","A","Patient1","KRAS","G12V"), ## KRAS G12V
                            c("chr12","25380275","25380275","T","A","Patient1","KRAS","Q61H"), ## KRAS Q61H
                            c("chr12","25380276","25380276","T","A","Patient1","KRAS","Q61L"), ## KRAS Q61L
                            c("chr12","25380276","25380276","T","C","Patient1","KRAS","Q61R"), ## KRAS Q61R
                            c("chr1","115256530","115256530","G","T","Patient1","NRAS","Q61K"), ## NRAS Q61K
                            c("chr1","115256528","115256528","T","G","Patient1","NRAS","Q61H"), ## NRAS Q61H
                            c("chr1","115256529","115256529","T","A","Patient1","NRAS","Q61L"), ## NRAS Q61L
                            c("chr1","115256529","115256529","T","C","Patient1","NRAS","Q61R"), ## NRAS Q61R
                            c("chr12","25398284","25398284","C","A","Patient2","KRAS","G12V"), ## KRAS G12V
                            c("chr12","25398285","25398285","C","G","Patient4","KRAS","G12R"), ## KRAS G12R
                            c("chr12","25380275","25380275","T","A","Patient4","KRAS","Q61H"), ## KRAS Q61H
                            c("chr12","25398284","25398284","C","G","Patient7","KRAS","G12A"), ## KRAS G12A
                            c("chr12","25398284","25398284","C","A","Patient7","KRAS","G12V"), ## KRAS G12V
                            c("chr1","115256528","115256528","T","A","Patient7","NRAS","Q61H"), ## NRAS Q61H
                            c("chr1","115256530","115256530","G","T","Patient7","NRAS","Q61K"), ## NRAS Q61K
                            c("chr1","115256529","115256529","T","A","Patient7","NRAS","Q61L"), ## NRAS Q61L
                            c("chr12","25398284","25398284","C","T","Patient9","KRAS","G12D"), ## KRAS G12D
                            c("chr1","115256530","115256530","G","T","Patient9","NRAS","Q61K"), ## NRAS Q61K
                            c("chr12","25398284","25398284","C","A","Patient10","KRAS","G12V"), ## KRAS G12V
                            c("chr12","25380275","25380275","T","A","Patient10","KRAS","Q61H"), ## KRAS Q61H
                            c("chr12","25380275","25380275","T","G","Patient10","KRAS","Q61H"), ## KRAS Q61H
                            c("chr12","25398284","25398284","C","G","Patient12","KRAS","G12A"), ## KRAS G12A
                            c("chr12","25398285","25398285","C","A","Patient12","KRAS","G12C"), ## KRAS G12C
                            c("chr12","25380275","25380275","T","G","Patient12","KRAS","Q61H"), ## KRAS Q61H
                            c("chr12","25398285","25398285","C","G","PatientBARD101","KRAS","G12R"), ## KRAS G12R
                            c("chr12","25380275","25380275","T","A","PatientBARD101","KRAS","Q61H"), ## KRAS Q61H
                            c("chr1","115256529","115256529","T","C","PatientBARD101","NRAS","Q61R"), ## NRAS Q61R
                            c("chr12","25398285","25398285","C","G","PatientBARD102","KRAS","G12R"), ## KRAS G12R
                            c("chr12","25398285","25398285","C","G","PatientBARD102","KRAS","G12R"), ## KRAS G12R
                            c("chr12","25380277","25380277","G","C","PatientBARD102","KRAS","Q61E"), ## KRAS Q61E
                            c("chr12","25398284","25398284","C","T","PatientBARD103","KRAS","G12D"), ## KRAS G12D
                            c("chr12","25398284","25398284","C","A","PatientBARD103","KRAS","G12V"), ## KRAS G12V
                            c("chr12","25398285","25398285","C","A","CRC188","KRAS","G12C"), ## KRAS G12C
                            c("chr12","25380275","25380275","T","A","CRC189","KRAS","Q61H"), ## KRAS Q61H
                            c("chr12","25380275","25380275","T","G","CRC189","KRAS","Q61H"), ## KRAS Q61H
                            c("chr12","25398284","25398284","C","G","CRC190","KRAS","G12A"), ## KRAS G12A
                            c("chr12","25380275","25380275","T","G","CRC190","KRAS","Q61H"), ## KRAS Q61H
                            c("chr12","25380276","25380276","T","C","CRC190","KRAS","Q61R"), ## KRAS Q61R
                            c("chr1","115256529","115256529","T","A","CRC190","NRAS","Q61L"), ## NRAS Q61L
                            c("chr1","115256529","115256529","T","C","CRC190","NRAS","Q61R"), ## NRAS Q61R
                            c("chr12","25398284","25398284","C","T","CRC191","KRAS","G12D"), ## KRAS G12D
                            c("chr12","25398285","25398285","C","G","CRC191","KRAS","G12R"), ## KRAS G12R
                            c("chr12","25398284","25398284","C","A","CRC191","KRAS","G12V"), ## KRAS G12V
                            c("chr12","25380275","25380275","T","G","CRC191","KRAS","Q61H")) ## KRAS Q61H

```

```{r,eval=T,echo=T,fig.width=5, fig.height=5, fig.align = "center"}

## merge the mutation data and ensure we are only considering KRAS/NRAS mutations
KRASNRAS_KnownMuts<-rbind(Prospect1_KnownMuts,Prospect2_KnownMuts,Bettegowda_KnownMuts); KRASNRAS_KnownMuts=KRASNRAS_KnownMuts[which(KRASNRAS_KnownMuts[,7]%in%c("KRAS","NRAS")),];

## Total number of mutations considered
All.Mut.Prop<-dim(KRASNRAS_KnownMuts)[1]

## What proportion of those mutations are Q61H
Q61.Prop<-dim(KRASNRAS_KnownMuts[which(KRASNRAS_KnownMuts[,8]=="Q61H"),])[1]/All.Mut.Prop

## What proportion of those mutations are not Q61H
notQ61.Prop<-dim(KRASNRAS_KnownMuts[which(KRASNRAS_KnownMuts[,8]!="Q61H"),])[1]/All.Mut.Prop

## make barplot
mids<-barplot(c(notQ61.Prop,Q61.Prop),col=c("red","blue"),ylim=c(0,1),ylab="Proportion of KRAS/NRAS codon 12/13/61")

axis(1,c("G12 G13 Q61 (Not Q61H)","Q61H"),at=mids)

## add fold change
text(x=mids[1]+(mids[2]-mids[1])/2,y=0.9,labels=paste0(round(notQ61.Prop/Q61.Prop,2),"-fold fewer Q61H generated"))

```

# Figure S7C

## Set parameters on clinical data 

Clinical parmeters for the model derived from the 12 BL prolonged benefit tumors.

```{r clinical data,echo=T,eval=T}

## age (years) at diagnosis of mets
AGE_AT_DIAGNOSIS=68.4

## median age (years) at BL biopsy
AGE_AT_BIOPSY=71.1

## median time (years) since diagnosis of mets
DIAGNOSIS_TO_BIOPSY=AGE_AT_BIOPSY-AGE_AT_DIAGNOSIS

## 2x the time from diagnosis as tumours are likely there for 6 months to a few years before diagnosed
AGE_AT_TUMOUR_INITIATION=AGE_AT_BIOPSY-(2*DIAGNOSIS_TO_BIOPSY)

```

## Define which hotspot mutations to consider

These are all hotspots shown in Figure 3A. Please refer to the manuscript for more information.

```{r Define hotspot mutations, echo=T,eval=T}

##KRAS/NRAS mutation in Figure 3A
hotspot.contexts=rbind(
  c("A[C>G]C","KRAS G12A"),
  c("C[C>A]A","KRAS G12C"),
  c("A[C>T]C","KRAS G12D"),
  c("C[C>G]A","KRAS G12R"),
  c("C[C>T]A","KRAS G12S"),
  c("A[C>A]C","KRAS G12V"),
  c("G[C>T]C","KRAS G13D"),
  c("C[C>A]A","KRAS G13C"),
  c("T[C>G]A","KRAS Q61E"),
  c("C[T>A]T","KRAS Q61H (T>A)"),
  c("C[T>G]T","KRAS Q61H (T>G)"),
  c("T[C>A]A","KRAS Q61K"),
  c("T[T>A]G","KRAS Q61L"),
  c("T[T>G]G","KRAS Q61P"),
  c("T[T>C]G","KRAS Q61R"),
  c("A[C>G]C","NRAS G12A"),
  c("C[C>A]T","NRAS G12C"),
  c("A[C>T]C","NRAS G12D"),
  c("C[C>T]T","NRAS G12S"),
  c("A[C>A]C","NRAS G12V"),
  c("A[C>T]C","NRAS G13D"),
  c("C[C>G]A","NRAS G13R"),
  c("C[T>A]T","NRAS Q61H (T>A)"),
  c("C[T>G]T","NRAS Q61H (T>G)"),
  c("T[T>A]G","NRAS Q61L"),
  c("A[C>A]A","NRAS Q61K"),
  c("T[T>C]G","NRAS Q61R"))

```

## Constant and accelerating mutation accumulation

Here we look to model the 'permanant' signatures (i.e. those that have accumulated prior to chemotherapy treatment). This set will always include the clock-like signatures (SBS1, SBS5, SBS40). In addition, some models will include the HR (SBS3), MMR (SBS6, SBS15) and SBS17a signatures due to the uncertainty of when these mutational processes start. 

Much of the following script is written using functions to allow for easy re-running after varying model parameters.

For the permant signatures we assume a constant accumulation of mutations from birth to diagnosis (or more generally 'acceleration.start'). This is followed by a period of accelerated mutation accumulation which we will vary. A constant of the model is that the total mutations will always equal the same number. So the constant rate and the acceleration part will need to adjust to that number. The conversion is as follows... 

Acceleration period...
$$
GROWTH_{YRS} = BIOPSY_{YRS} - DIAGNOSIS_{YRS}
$$

Total mutations are split into a constant portion and an acceleration portion...
$$
MUTATIONS_{TOTAL} = MUTATIONS_{CONSTANT} + MUTATIONS_{ACCELERATION}
$$

Rearrange the formula as follows...
$$
MUTATIONS_{TOTAL} = (RATE \times DIAGNOSIS_{YRS}) + (RATE \times GROWTH_{YRS} \times ACCELERATION)
$$

$$
MUTATIONS_{TOTAL} = RATE [DIAGNOSIS_{YRS} + GROWTH_{YRS} \times ACCELERATION]
$$

$$
RATE = MUTATIONS_{TOTAL} / [ DIAGNOSIS_{YRS} + ACCELERATION \times GROWTH_{YRS}]
$$
```{r,echo=T,eval=T}
MutationRateAdjustment<-function(acceleration,constant.signatures,growth_years){
  
  rate<-Signature.AssignedMutations[constant.signatures,]/(acceleration.start+growth_years*acceleration)

  return(rate)
}
```

We can then split the clock like mutation load into constant and accelerating portions...

```{r split clock like mutation load into constant and accelerating portions,echo=T,eval=T}

raw.mutation.split<-function(constant.signatures,acceleration.start,acceleration.end){

  growth_years=acceleration.end-acceleration.start
  
  MutationSplits<-matrix(NA,length(accelerations),3); rownames(MutationSplits)<-paste0("x",accelerations); colnames(MutationSplits)<-c("Constant","Accelerating","Total")
  
  for(j in 1:length(accelerations)){ 
    
    acceleration<-accelerations[j]
    
    rate<-MutationRateAdjustment(acceleration=accelerations[j],constant.signatures,growth_years=acceleration.end-acceleration.start)
    
    MutationSplits[j,]['Constant']<-sum(acceleration.start*rate)
    MutationSplits[j,]['Accelerating']<-sum(growth_years*rate*acceleration)
    
  }
  
  MutationSplits[,'Total']<-MutationSplits[,'Constant']+MutationSplits[,'Accelerating']
  
  MutationSplits<-round(MutationSplits,0)
  
  return(MutationSplits)

}

```

```{r acceleration plot,echo=T,eval=T, fig.align = "center"}

acceleration.plot<-function(MutationSplits,acceleration.start,acceleration.end){

## empty plot to add the trejectory
plot(1, type="n", xlim=c(0,80),ylim=c(0,2000),xlab="Age (years)",ylab="Total Mutations",bty='n')

## add the acceleration boundary years
axis(1,at=c(acceleration.start,acceleration.end))

## add acceleration boundaries
abline(v=acceleration.start,lty=2,col="grey")
abline(v=acceleration.end,lty=2,col="grey")

## colours for each rate
colours=c("black","purple","green")

## add legend linking colour to rate
legend("topleft",col=colours,legend=paste0("x",accelerations),bty='n',lty=1)

for(j in 1:length(accelerations)){
  
  segments(x0=0,x1=acceleration.start,y0=0,y1=MutationSplits[,'Constant'][j],col=colours[j])
  
  segments(x0=acceleration.start,x1=acceleration.end,y0=MutationSplits[,'Constant'][j],y1=MutationSplits[,'Total'][j],col=colours[j])
  
}
}

```

```{r demo plot,echo=T,eval=T,fig.width=5, fig.height=5, fig.align = "center"}

## set model parameters
accelerations=c(1,5,10)
constant.signatures=c("SBS1","SBS5","SBS40")
acceleration.start=AGE_AT_DIAGNOSIS
acceleration.end=AGE_AT_BIOPSY

## generate the matrix of mutation splits
MutationSplits<-raw.mutation.split(constant.signatures=constant.signatures,acceleration.start=acceleration.start,acceleration.end=acceleration.end)

MutationSplits

## plot the trajectories of mutation accumulation
acceleration.plot(MutationSplits=MutationSplits,acceleration.start=acceleration.start,acceleration.end=acceleration.end)

```

# Figure S7D

## Adjusted signature exposures

```{r generate adjusted signature exposure matrix,echo=T,eval=T}

adjusted.exposures<-function(constant.signatures,temp.signatures,acceleration.start,acceleration.end){
  
  growth_years=acceleration.end-acceleration.start
    
  acceleration.adjusted=list(); for(j in 1:length(accelerations)){
  
    acceleration<-accelerations[j]
    
    rate<-MutationRateAdjustment(acceleration=acceleration,constant.signatures=constant.signatures,growth_years=growth_years)
    
    ## temporary signatures
    temp.sigs=Signature.AssignedMutations[temp.signatures,]
      
    ## acceleration adjusted signatures
    acceleration.adjusted.sigs=rate*growth_years*acceleration
    
    acceleration.adjusted[[paste0("x",accelerations[j])]]<-rbind(acceleration.adjusted.sigs,temp.sigs)[selected.signatures,]

  }
  return(acceleration.adjusted)
}

```

```{r demo adjusted exposures,echo=T,eval=T}

constant.signatures=c("SBS1","SBS5","SBS40")
temp.signatures=c("SBS3","SBS6","SBS15","SBS17a","SBS17b","SBS35")
acceleration.start=AGE_AT_DIAGNOSIS
acceleration.end=AGE_AT_BIOPSY

weighted.signatures<-adjusted.exposures(constant.signatures=constant.signatures,
                   temp.signatures=temp.signatures,
                   acceleration.start=acceleration.start,
                   acceleration.end=acceleration.end)

weighted.signatures$x10

```

```{r weight signatures,echo=T,eval=T}

constrain.weighted.signatures<-function(props,constrained.sigs){

  if(length(which(constrained.sigs!="ALL")>0)){
  constrained.sigs<-unlist(strsplit(constrained.sigs,"[|]"))
  props[constrained.sigs]<-0
  }

  constrained.weighted.sigs<-t(as.matrix(props*t(Canonical_Signatures_selected_ExomeScaled)));

return(constrained.weighted.sigs)
}

```

```{r}

weighted.signatures.props<-lapply(weighted.signatures,function(x) rowSums(x)/sum(x))

constrain=c("ALL","SBS17b","SBS35","SBS17b|SBS35")

weighted.signatures=list(); for(i in 1:length(constrain)) weighted.signatures[[constrain[i]]]<-lapply(weighted.signatures.props,function(x) constrain.weighted.signatures(x,constrain[i]))

weighted.signatures$`SBS17b|SBS35`$x1

```

```{r,echo=T,eval=T}

signature.influence<-function(Hotspot_Probability){

  Q61HProb<-lapply(Hotspot_Probability,function(x) sum(as.numeric(x[grep("Q61H",x[,2]),][,3])))
  
  OtherProb<-lapply(Hotspot_Probability,function(x) sum(as.numeric(x[grep("Q61H",x[,2],invert=T),][,3])))
  
  mutation.probability.table<-rbind((unlist(Q61HProb)/(unlist(Q61HProb)+unlist(OtherProb)))*100,
  (unlist(OtherProb)/(unlist(Q61HProb)+unlist(OtherProb)))*100)
  rownames(mutation.probability.table)<-c("Q61H","Other")
return(mutation.probability.table)
  
}

```

```{r}

Hotspot_Probabilities=list(); for(i in 1:length(constrain)) Hotspot_Probabilities[[constrain[i]]]<-lapply(weighted.signatures[[constrain[i]]],function(x) cbind(hotspot.contexts,rowSums(x)[hotspot.contexts[,1]]))

Hotspot_Probabilities$`SBS17b|SBS35`$x10

```

## Q61H versus other hotspots

```{r barplots of Q61 probability versus all other drivers,echo=T,eval=T}

barplot.Q61<-function(Q61Table,acceleration){

plot.vals<-as.table(rbind(Q61Table[['SBS17b|SBS35']],
Q61Table[['SBS35']],
Q61Table[['SBS17b']],
Q61Table[['ALL']])); rownames(plot.vals)<-c("noSBS17b.noSBS35.Q61H","noSBS17b.noSBS35.Other","noSBS17b.Q61H","noSBS17b.Other","noSBS35.Q61H","noSBS35.Other","ALL.Q61H","ALL.Other");

## constant acceleration
mat<-as.table(rbind(plot.vals[,paste0("x",acceleration)][grep("Q61H",rownames(plot.vals),invert=T)],plot.vals[,paste0("x",acceleration)][grep("Q61H",rownames(plot.vals))])); 

fold.changes<-apply(rbind(mat[1,]/mat[2,],mat[2,]/mat[1,]),2,function(x) max(x))

mids<-barplot(mat,beside=T,main=paste0("acceleration: x",acceleration),col=c("red","blue"), xaxt='n',ylim=c(0,100));

text(x=mids[1,]+(mids[2,]-mids[1,])/2,y=98,labels=round(fold.changes,2))

axis(1, at=mids[1,]+(mids[2,]-mids[1,])/2, labels=c("-","SBS17b","SBS35","All"),las=1)

}

```

```{r}


Q61versusRest<-lapply(Hotspot_Probabilities,function(x) signature.influence(x))

barplot.Q61(Q61versusRest,acceleration=10)

```

## Modelling Pipeline

```{r}

model.pipe<-function(){

par(mfrow=c(2,2))

MutationSplits<-raw.mutation.split(constant.signatures=constant.signatures,acceleration.start=acceleration.start,acceleration.end=acceleration.end)

acceleration.plot(MutationSplits=MutationSplits,acceleration.start=acceleration.start,acceleration.end=acceleration.end)

acceleration.adjusted.signatures<-adjusted.exposures(constant.signatures=constant.signatures,temp.signatures=temp.signatures,acceleration.start=acceleration.start,acceleration.end=acceleration.end)

proportional.sig.contribution<-lapply(acceleration.adjusted.signatures,function(x) rowSums(x)/sum(x))

constrain=c("ALL","SBS17b","SBS35","SBS17b|SBS35")

weighted.signatures=list(); for(i in 1:length(constrain)) weighted.signatures[[constrain[i]]]<-lapply(proportional.sig.contribution,function(x) constrain.weighted.signatures(x,constrain[i]))

Hotspot_Probabilities=list(); for(i in 1:length(constrain)) Hotspot_Probabilities[[constrain[i]]]<-lapply(weighted.signatures[[constrain[i]]],function(x) cbind(hotspot.contexts,rowSums(x)[hotspot.contexts[,1]]))

Q61versusRest<-lapply(Hotspot_Probabilities,function(x) signature.influence(x))

for(acceleration in accelerations) barplot.Q61(Q61versusRest,acceleration)
  
}

```

## Model 1

Permanant Signatures: SBS1, SBS3, SBS5, SBS6, SBS15, SBS17a, SBS40

Temporary Signatures: SBS17b, SBS35

Tumour Growth Period: 2.7 years

```{r,fig.width=10, fig.height=10, fig.align = "center"}

constant.signatures<-c('SBS1','SBS3','SBS5','SBS6','SBS15','SBS17a','SBS40')
temp.signatures<-c('SBS17b','SBS35')

## acceleration parameters to model
accelerations=c(1,5,10)

## define the start and end of the tumour growth period
acceleration.start=AGE_AT_DIAGNOSIS
acceleration.end=AGE_AT_BIOPSY

model.pipe()

```

## Model 2

Permanant Signatures: SBS1, SBS5, SBS17a, SBS40

Temporary Signatures: SBS3, SBS6, SBS15, SBS17b, SBS35

Tumour Growth Period: 2.7 years

```{r,fig.width=10, fig.height=10, fig.align = "center"}

## defiine permanat and temporary signatures
constant.signatures<-c('SBS1','SBS5','SBS17a','SBS40')
temp.signatures<-c('SBS3','SBS6','SBS15','SBS17b','SBS35')

## acceleration parameters to model
accelerations=c(1,5,10)

## define the start and end of the tumour growth period
acceleration.start=AGE_AT_DIAGNOSIS
acceleration.end=AGE_AT_BIOPSY

model.pipe()

```

## Model 3

Permanant Signatures: SBS1, SBS3, SBS5, SBS6, SBS15, SBS17a, SBS40

Temporary Signatures: SBS17b, SBS35

Tumour Growth Period: 5.4 years

```{r,fig.width=10, fig.height=10, fig.align = "center"}

## defiine permanat and temporary signatures
constant.signatures<-c('SBS1','SBS3','SBS5','SBS6','SBS15','SBS17a','SBS40')
temp.signatures<-c('SBS17b','SBS35')

## acceleration parameters to model
accelerations=c(1,5,10)

## define the start and end of the tumour growth period
acceleration.start=AGE_AT_TUMOUR_INITIATION
acceleration.end=AGE_AT_BIOPSY

model.pipe()

```

## Model 4

Permanant Signatures: SBS1, SBS5, SBS17a, SBS40

Temporary Signatures: SBS3, SBS6, SBS15, SBS17b, SBS35

Tumour Growth Period: 5.4 years

```{r,fig.width=10, fig.height=10, fig.align = "center"}

## defiine permanat and temporary signatures
constant.signatures<-c('SBS1','SBS5','SBS17a','SBS40')
temp.signatures<-c('SBS3','SBS6','SBS15','SBS17b','SBS35')

## acceleration parameters to model
accelerations=c(1,5,10)

## define the start and end of the tumour growth period
acceleration.start=AGE_AT_TUMOUR_INITIATION
acceleration.end=AGE_AT_BIOPSY

model.pipe()

```

# Session Information

```{r Session Info,echo=F,eval=T}
devtools::session_info()

# Session Workspace
save.image(file=".Rdata")
```